bftengine/src/bcstatetransfer/RangeValidationTree.hpp

```diff
@@ -0,0 +1,268 @@
+    }
+    std::string toString() const noexcept;
+
+    static constexpr size_t kNIDBitsPerLevel = 8;
+    static constexpr size_t kNIDBitsPerRVBIndex = ((sizeof(uint64_t) * 8) - kNIDBitsPerLevel);
+    static constexpr size_t kMaxLevels = ((0x1 << kNIDBitsPerLevel) - 1);
+    static constexpr uint64_t kRvbIdMask = (kMaxLevels) << kNIDBitsPerRVBIndex;
+    static constexpr uint64_t kRvbIndexMask = std::numeric_limits<uint64_t>::max() & (~kRvbIdMask);
+
+    // node location in the tree
+    const uint64_t level : kNIDBitsPerLevel;
+    const uint64_t rvb_index : kNIDBitsPerRVBIndex;
+
+    // node ID
+    const uint64_t id;
+  };
+
+  struct RVBNode {
+    RVBNode(uint64_t rvb_index, const char* data, size_t data_size);
+    RVBNode(uint8_t level, uint64_t rvb_index);
+    RVBNode(uint64_t node_id, char* val, size_t size);
+
+    bool isMinChild() { return info_.rvb_index % RVT_K == 1; }
+    bool isMaxChild() { return info_.rvb_index % RVT_K == 0; }
+    void logInfoVal(const std::string& prefix = "");
+    const std::shared_ptr<char[]> computeNodeInitialValue(NodeInfo& node_id, const char* data, size_t data_size);
+
+    static constexpr uint8_t kDefaultRVBLeafLevel{0};
+    NodeInfo info_;
+    NodeVal current_value_;
+    static uint32_t RVT_K;
+  };
+
+ public:
+#pragma pack(push, 1)
+  struct RVTMetadata {
+    uint64_t magic_num{0};
+    uint8_t version_num{0};
+    uint32_t RVT_K{0};
+    uint32_t fetch_range_size{0};
+    size_t value_size{0};
+    uint64_t total_nodes{0};
+    uint64_t root_node_id{0};
+
+    static void staticAssert() noexcept;
+  };
+#pragma pack(pop)
+
+  struct SerializedRVTNode {
+    uint64_t id{0};
+    size_t current_value_encoded_size{0};
+    uint16_t n_child{0};
+    uint64_t min_child_id{0};
+    uint64_t max_child_id{0};
+    uint64_t parent_id{0};
+
+    static void staticAssert() noexcept;
+  };
+
+  struct RVTNode;
+  using RVBNodePtr = std::shared_ptr<RVBNode>;
+  using RVTNodePtr = std::shared_ptr<RVTNode>;
+  struct RVTNode : public RVBNode {
+    RVTNode(const RVBNodePtr& node);
+    RVTNode(const RVTNodePtr& node);
+    RVTNode(SerializedRVTNode& node, char* cur_val_ptr, size_t cur_value_size);
+    static RVTNodePtr createFromSerialized(std::istringstream& is);
+
+    void addValue(const NodeVal& nvalue);
+    void substructValue(const NodeVal& nvalue);
+    std::ostringstream serialize() const;
+
+    static constexpr uint8_t kDefaultRVTLeafLevel = 1;
+    uint16_t n_child{0};
+    uint64_t min_child_id{0};      // Minimal actual child id
+    uint64_t max_child_id{0};      // Maximal possible child id. The max actual is min_child_id + n_child.
+    uint64_t parent_id{0};         // for root - will be 0
+    const NodeVal initial_value_;  // We need to keep this value to validate node's current value
+  };
+
+ public:
+  // validation functions
+  static uint64_t pow_int(uint64_t base, uint64_t exp) noexcept;
```

Comment:
Maybe this function can be renamed to `pow_uint`, since it only operates on unsigned integers?